docs/docs/reference-react-component.md

```diff
@@ -230,34 +230,59 @@ componentWillUnmount()
 ### `setState()`
 
 ```javascript
-setState(nextState, callback)
+setState(updater, [callback])
 ```
 
-Performs a shallow merge of nextState into current state. This is the primary method you use to trigger UI updates from event handlers and server request callbacks.
+`setState()` is an asynchronous method which enqueues changes to be made to the state during the next update cycle. This is the primary method you use to trigger updates in response to event handlers, server responses, etc...
 
-The first argument can be an object (containing zero or more keys to update) or a function (of state and props) that returns an object containing keys to update.
+`setState()` does not immediately mutate `this.state` but creates a pending state transition. Accessing `this.state` after calling this method can potentially return the previous state, rather than the state after enqueued updates have been applied. This is a common source of bugs in React applications.
```

Comment:
Maybe 

>`setState()` does not immediately mutate `this.state`. Instead, React mutates `this.state` when it applies the enqueued update, and this may happen later than the `setState()` call. This makes reading `this.state` right after calling `setState()` a potential pitfall.

?